### PR TITLE
#12 주문 생성

### LINKED ISSUE #12: 주문 구현

### PR DESCRIPTION
**Pull request by bgpark82**
## 개요
주문 생성 기능

## 이슈 티켓
#12 주문 구현

## 작업 내용
- [x] 주문 생성 Mapper, Service, Controller 생성
- [x] 배달 생성 Mapper 생성
- [x] JsonMapper 생성
- [x] 메뉴 옵션 조회, 가격 게산 로직을 메뉴로 이동

### CODE REVIEW

File: src/main/java/com/flab/commerce/domain/order/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+@RequiredArgsConstructor
+public class OrderController {
+
+  private final OrderService orderService;
+
+  @PostMapping("/orders")
+  public ResponseEntity<Void> save(@Valid @RequestBody OrderSaveDto saveDto) {
+    orderService.save(saveDto);
+    return ResponseEntity.ok().build();
+  }
```
Comment (f-lab-jason) [inline src/main/java/com/flab/commerce/domain/order/OrderController.java]: validation후의 처리는 따로 하지않나요?

File: src/main/java/com/flab/commerce/domain/price/PriceMapper.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.flab.commerce.domain.price;
+
+import org.apache.ibatis.annotations.Mapper;
+
+@Mapper
+public interface PriceMapper {
+
+  int save(Price newPay);
```
Comment (f-lab-jason) [inline src/main/java/com/flab/commerce/domain/price/PriceMapper.java]: 'pay'라는 이름을 쓰는 이유가 있나요?

File: src/main/java/com/flab/commerce/util/JsonMapper.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.flab.commerce.exception.BadParseException;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class JsonMapper<T> {
```
Comment (f-lab-jason) [inline src/main/java/com/flab/commerce/util/JsonMapper.java]: wrapper class를 따로 생성하신 이유가 있나요?
BadParseException을 발생시키기 위한거라면 ObjectMapper가 생성하는 exception을 그대로 써도 무방해보이네요.

File: src/main/java/com/flab/commerce/domain/order/OrderSaveDto.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+@AllArgsConstructor
+public class OrderSaveDto {
+
+  @NotNull
+  private Long menuId;
+
+  @NotNull
+  private Long userId;
+
+  private int amount;
```
Comment (f-lab-jason) [inline src/main/java/com/flab/commerce/domain/order/OrderSaveDto.java:27]: (nit) 이부분도 양수 validation이 필요하겠네요.

**Review by f-lab-jason — APPROVED** [review]
1. Description도 작성해주세요.
2. 빠른 구현을 위해 생략한 부분이 있다면 같이 명시해주시면 좋겠습니다. (validation이나 테스트 케이스 등)

리뷰 범위에 대해서 세션에서 한 번 얘기해볼 필요가 있겠네요 🙂

File: src/main/java/com/flab/commerce/domain/order/OrderController.java (hunk 1 above)
Comment (bgpark82) [reply src/main/java/com/flab/commerce/domain/order/OrderController.java]: 혹시 어떤 처리를 말씀하시는 걸까요?

제가 잘 몰라서 그럴 수 있는데 어떤 부분들이 필요한지 알려주시면 감사하겠습니다🙏

File: src/main/java/com/flab/commerce/domain/price/PriceMapper.java (hunk 2 above)
Comment (bgpark82) [reply src/main/java/com/flab/commerce/domain/price/PriceMapper.java]: Price라는 네이밍 사용전에 Pay로 개발했었는데 남아 있던 것 같습니다.
수정해야겠습니다

File: src/main/java/com/flab/commerce/util/JsonMapper.java (hunk 3 above)
Comment (bgpark82) [reply src/main/java/com/flab/commerce/util/JsonMapper.java]: 객체를 String으로 파싱하는 책임을 가진 클래스를 따로 생성했습니다.
나중에 String을 메뉴 옵션 객체로 변환할 텐데 그때 이 클래스만 변경하면 될 것 같아 생성했었습니다.
말씀하신대로 exception은 ObjectMapper가 생성하는 예외를 그대로 써도 무방했지만 한 클래스에서 관리가 되서 겸사겸사 클래스로 빼게 되었습니다

좀 성급하게 클래스를 나눈 걸까요?

File: src/main/java/com/flab/commerce/domain/order/OrderSaveDto.java (hunk 4 above)
Comment (bgpark82) [reply src/main/java/com/flab/commerce/domain/order/OrderSaveDto.java:27]: 네 Validation과 테스트 추가했습니다

**Review by bgpark82 — COMMENTED** [review]
## 추가 작업 내용
- [x] Description 작성
- [x] `amount` 양수 Validation 추가 
- [x] 파라미터 명 변경 (`newPay` -> `newPrice`)


## 추가 질문

추가적인 질문을 드리고 싶습니다.

1. **Description에 보통 들어가는 내용이 어떤 것이 있을까요?**

지금 회사에서는 개요, 문서, 변경 내용을 작성합니다. 
혹시 추가적으로 필요하다 생각하시는 부분이나 중요하다고 생각하는 부분이 있는지
또한 회사에서는 어떤 내용을 pr description에 사용하고 계신지 궁금합니다. 

2. **빠른 구현을 위해 생략한 부분**

**validation**은 아래 2경우를 검증했다 생각합니다.
- dto :  @Valid 적용 및 Controller 테스트로 에지케이스 검증 (`OrderControllerTest`)
- 로직 : service 레이어에서 로직의 validation과 service 혹은 domain 테스트 (`OrderServiceTest`, `MenuTest`)

제가 잘 몰라서 그럴 것 같은데 혹시 어떤 부분을 더 Validation 해야되고 어떤 테스트가 더 필요할지 궁금합니다!

리뷰 범위는 말씀대로 수업시간에 이야기 나눠보면 좋을 것 같습니다 🙂

File: src/main/java/com/flab/commerce/domain/order/OrderController.java (hunk 1 above)
Comment (f-lab-jason) [reply src/main/java/com/flab/commerce/domain/order/OrderController.java:20]: 앗 네, `@Valid` annotation이 어떻게 동작하는지 찾아보시면 좋겠습니다. validation 후의 에러 처리는 `BindingResult`를 받아서 따로 해주셔야합니다.

Comment (f-lab-jason) [reply src/main/java/com/flab/commerce/domain/order/OrderController.java]: 음 아닙니다 위 코멘트는 무시해주세요 😅 
`ResponseEntityExceptionHandler`를 확장해서 처리하는걸 생각했는데 spring에서 주는 error를 그대로 써도 크게 상관없겠습니다.

File: src/main/java/com/flab/commerce/util/JsonMapper.java (hunk 3 above)
Comment (f-lab-jason) [reply src/main/java/com/flab/commerce/util/JsonMapper.java]: 아 질문드린건 그냥 `ObjectMapper`를 주입받아서 단순히 `mapper.writeValueAsString(obj)`를 호출하는것과 어떤 차이가 있는지 궁금해서 여쭤봤어요.
String을 메뉴 옵션 객체로 변환하는것도 `mapper.readValue(json, ${class})`외에 다른 작업이 필요한가요?

**Comment by f-lab-jason** [comment]
1. Description은 사실 회사마다 사용하는 포맷이 다르긴해서 정답은 없을 것 같고 작업내용을 잘 표현할 수 있는 정도면 좋겠습니다. 참고로 현재 저희 회사에서는 PR마다 아래와 같이 Notify, Summary, Motivation, Test plan, Rollout/monitoring/revert plan으로 나눠진 템플릿을 작성하고 있습니다.
  사실 Description은 전적으로 리뷰어를 위한거니 가급적 리뷰어가 편할 수 있도록 굉장히 자세하게 적는 편이긴합니다. (보통 리뷰어는 아무것도 모른다는 가정하에 작성하는거죠.)
```
<!-- If this branch is in-progress, start the title with [wip]. 
CIBot will prevent this branch from being merged until the title is edited to remove the leading [wip]. -->
### Notify
r? @stripe-internal/datastore-reviewers 

<!-- Assign codeowners as reviewers by commenting `r?` on a single line. 
See go/code-review for more guidance on the code review process. -->

### Summary
<!-- What does the code do? What have you changed? 
If this is a visual change consider including a screenshot or gif. 
See go/screencap for tips/tools. -->

### Motivation
<!-- Why are you making this change? This can be a link to a Jira task. -->

### Test plan
<!-- How did you test this change? What were you unable to test? Please include additional context, 
e.g. were you able to cover failures and edge cases? 
Reference automated tests or describe a manual test plan and confirm the outcome.
Please keep the frontpage test, our auditors (who review a random sample of our PRs), 
and your reviewer in mind. In cases where you are unable to test your changes, or it is not appropriate, 
please leave both boxes unchecked. -->

### Rollout/monitoring/revert plan
<!--
  What services will you deploy? Include all services which could be impacted by the change. 
Hint: CIBot can tag a PR with services using the `s: service` syntax.

Are there any post-deploy steps (e.g. run migration or enable feature flag)? 
If this change requires special attention to rollout, 
what will you monitor to make sure the deployment is proceeding as expected? 
If the change needs to be rolled back or reverted, are there any additional steps that need to be taken? 
If this change needs to be manually deployed or deployed with unusual tooling, 
please list the specific steps required to deploy — and if it becomes necessary — roll back this change.

Changes in the critical payments path must be decoupled from the deploy and have a monitoring plan. 
This can usually be achieved with a feature flag or gate. 
See go/code-reviewer-guide and go/safe-payflows-rollouts for details.
-->
```

2. 테스트와 리뷰범위에 대해서는 세션시간에 잠깐 더 얘기하도록 해요 🙂

File: src/main/java/com/flab/commerce/domain/order/OrderController.java (hunk 1 above)
Comment (bgpark82) [reply src/main/java/com/flab/commerce/domain/order/OrderController.java]: `ResponseEntityExceptionHandler`와 `@RestControllerAdvice`로 전역 예외 처리를 하면 좋을 것 같네요!
일단 `MethodArgumentNotValidException`에 대해서만 예외처리를 이렇게 하도록 하겠습니다

File: src/main/java/com/flab/commerce/util/JsonMapper.java (hunk 3 above)
Comment (bgpark82) [reply src/main/java/com/flab/commerce/util/JsonMapper.java]: 아 네 `mapper.writeValueAsString(obj)`와 단순 차이는 없습니다.
말씀대로 `mapper.readValue(json, ${class})` 외에는 다른 작업이 필요다고 생각됩니다.

ObjectMapper를 사용할 때마다 try-catch로 예외처리 하는 부분이 반복될 것 같아 클래스로 빼게 되었습니다
제가 잘 모르는 걸 수도 있는데 JsonProcessingException를 throws로 넘겨 버리면 해당 메소드를 사용하는 메소드들이 throws로 계속 전파되서 해당 클래스 내부에서 처리하는게 좋다고 생각했습니다

이런 경우에는 그냥 throws로 넘기고 Objectmapper가 생성하는 exception을 그대로 사용하는게 좋을까요?

**Review by bgpark82 — COMMENTED** [review]
## 추가 작업 내용
- [x] MethodArgumentNotValidException 전역 예외 처리

피드백 감사드립니다! 
다음 PR부터는 Description에 이런 내용도 있었으면 좋겠다고 피드백 주시면 감사할 것 같습니다
말씀대로 테스트와 리뷰 범위는 수업시간에 이야기 해보는게 좋을 것 같습니다 :)

File: src/main/java/com/flab/commerce/util/JsonMapper.java (hunk 3 above)
Comment (f-lab-jason) [reply src/main/java/com/flab/commerce/util/JsonMapper.java]: 아 네, `@ControllerAdvice`와 `@ExceptionHandler`를 조합해서 `JsonProcessingException`을 처리하는 방법을 찾아보시면 좋겠습니다.

Comment (bgpark82) [reply src/main/java/com/flab/commerce/util/JsonMapper.java]: 네 감사합니다!
JsonMapper 제거하고 JsonProcessingException으로 처리했습니다
한번 확인 부탁드리겠습니다

**Review by bgpark82 — COMMENTED** [review]
- [x] JsonProcessingException 예외처리 확장
- [x] JsonMapper 제거

**Review by f-lab-jason — APPROVED** [review]
🚢
